specs/ledger/hs/src/Ledger/Update.hs

```diff
@@ -1,29 +1,248 @@
   -- ^ Maximum (abstract) block header size in words.
-  , _dLiveness :: !SlotCount
+  , _maxTxSz :: Natural
+  -- ^ Maximum (abstract) transaction size in words.
+  , _maxPropSz :: Natural
+  -- ^ Maximum (abstract) update proposal size in words.
+  , _dLiveness :: Core.SlotCount
   -- ^ Delegation liveness parameter: number of slots it takes a delegation
   -- certificate to take effect.
-  , _bkSgnCntW :: !Int
+  , _bkSgnCntW :: Int
   -- ^ Size of the moving window to count signatures.
-  , _bkSgnCntT :: !Double
+  , _bkSgnCntT :: Double
   -- ^ Fraction [0, 1] of the blocks that can be signed by any given key in a
   -- window of lenght '_bkSgnCntW'. This value will be typically between 1/5
   -- and 1/4.
-  , _bkSlotsPerEpoch :: !SlotCount
+  , _bkSlotsPerEpoch :: Core.SlotCount
+  -- ^ Number of slots in an epoch
+  , _upTtl :: Core.SlotCount
+  -- ^ Update proposal TTL in slots
+  , _scriptVersion :: Natural
+  -- ^ Script version
   } deriving (Eq, Show)
 
 makeLenses ''PParams
+
+newtype UpId = UpId Int
+  deriving (Eq, Ord, Show)
+
+-- | Protocol version
+data ProtVer = ProtVer
+  { _pvMaj :: Natural
+  , _pvMin :: Natural
+  , _pvAlt :: Natural
+  } deriving (Eq, Show)
+
+makeLenses ''ProtVer
+
+newtype ApName = ApName String
+  deriving (Eq, Ord, Show)
+
+-- | Application version
+newtype ApVer = ApVer Natural
+  deriving (Eq, Num, Show)
+
+data SwVer = SwVer
+  { _svName :: ApName
+  , _svVer :: ApVer
+  } deriving (Eq, Show, Generic)
+
+makeLenses ''SwVer
+
+-- | Update proposal
+data UProp = UProp
+  { _upId :: UpId
+  , _upIssuer :: Core.VKey
+  , _upParams :: PParams
+  , _upPV :: ProtVer
+  , _upSwVer :: SwVer
+  , _upSig :: Core.Sig UProp
+  }
+
+makeLenses ''UProp
+
+inRange :: Ord a => a -> a -> a -> Bool
+inRange l u x = l <= x && x <= u
+
+-- | Test if a pair is present in a map.
+inMap :: (Ord k, Eq v) => k -> v -> Map k v -> Bool
+inMap k v m = case Map.lookup k m of
+  Just x | x == v -> True
+  _ -> False
+
+(==>) :: Bool -> Bool -> Bool
+a ==> b = not a || (a && b)
+infix 1 ==>
+
+pvCanFollow
+  :: ProtVer
+  -> ProtVer
+  -> Bool
+pvCanFollow (ProtVer mjn min an) (ProtVer mjp mip ap)
+  = (mjp, mip, ap) < (mjn, min, an)
+  && (inRange 0 1 (mjn - mjp))
+  && ((mjp == mjn) ==> (mip +1 == min))
+  && ((mjp +1 == mjn) ==> (min == 0))
+
+-- | Check whether an update proposal marks a valid update
+--
+--   TODO At the moment we don't check size in here - should wel?
+canUpdate
+  :: PParams
+  -> UProp
+  -> Bool
+canUpdate pps prop =
+  (prop ^. upParams . maxBkSz <= 2 * pps ^. maxBkSz)
+  && (prop ^. upParams . maxBkSz >= prop ^. upParams . maxTxSz)
+  && (inRange 0 1 $ prop ^. upParams . scriptVersion - pps ^. scriptVersion)
+
+svCanFollow
+  :: Map ApName (ApVer, Core.Slot)
+  -> (ApName, ApVer)
+  -> Bool
+svCanFollow avs (an,av) =
+    (case Map.lookup an avs of
+      Nothing -> True
+      Just (x,_) -> av == x + 1
+    ) && (an `Set.notMember` Map.keysSet avs ==> av == ApVer 0)
+  where
+
+data UPSVV
```

Comment:
`UPSVV` stands for "Update Proposal Software Version Validation". Do you think it'd make sense to add it as a comment?